metrics/core/historical_types.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package core
+
+import (
+	"fmt"
+	"time"
+)
+
+type AggregationType string
+
+var (
+	AggregationTypeAverage AggregationType      = "average"
+	AggregationTypeMaximum AggregationType      = "max"
+	AggregationTypeMinimum AggregationType      = "min"
+	AggregationTypeMedian AggregationType       = "median"
+	AggregationTypeCount AggregationType        = "count"
+	AggregationTypePercentile50 AggregationType = "50-perc"
+	AggregationTypePercentile95 AggregationType =  "95-perc"
+	AggregationTypePercentile99 AggregationType = "99-perc"
+)
+
+// MultiTypedAggregations is the list of aggregations that can be either float or int
+var MultiTypedAggregations = []AggregationType{
+	AggregationTypeAverage,
+	AggregationTypeMaximum,
+	AggregationTypeMinimum,
+	AggregationTypeMedian,
+	AggregationTypePercentile50,
+	AggregationTypePercentile95,
+	AggregationTypePercentile99,
+}
+
+// AllAggregations is the set of all supported aggregations
+var AllAggregations = map[AggregationType]bool{
+	AggregationTypeAverage:      true,
+	AggregationTypeMaximum:      true,
+	AggregationTypeMinimum:      true,
+	AggregationTypeMedian:       true,
+	AggregationTypePercentile50: true,
+	AggregationTypePercentile95: true,
+	AggregationTypePercentile99: true,
+	AggregationTypeCount:        true,
+}
+
+// TimestampedMetricValue is a metric value with an associated timestamp
+type TimestampedMetricValue struct {
+	MetricValue
+	Timestamp time.Time
+}
+
+// AggregationValue is a description of aggregated MetricValues over time
+type AggregationValue struct {
+	Count *uint64
+
+	Aggregations map[AggregationType]MetricValue
+}
+
+// TimestampedAggregationValue is an aggregation value with an associated timestamp
+// and bucket size
+type TimestampedAggregationValue struct {
+	Timestamp  time.Time
```

Comment:
Is this the end time of the bucket, or start time (for Hawkular we return both timestamps to the user) ? Or user definable as long as it's consistent?